src/deploy/functions/containerCleaner.ts

```diff
@@ -0,0 +1,160 @@
+  "us-central2": "us",
+  "us-east1": "us",
+  "us-east4": "us",
+  "northamerica-northeast1": "us",
+  "southamerica-east1": "us",
+  "europe-west1": "eu",
+  "europe-west2": "eu",
+  "europe-west3": "eu",
+  "europe-west5": "eu",
+  "europe-west6": "eu",
+  "europe-central2": "eu",
+  "asia-east1": "asia",
+  "asia-east2": "asia",
+  "asia-northeast1": "asia",
+  "asia-northeast2": "asia",
+  "asia-northeast3": "asia",
+  "asia-south1": "asia",
+  "asia-southeast2": "asia",
+  "australia-southeast1": "asia",
+};
+
+export async function cleanupBuildImages(functions: backend.FunctionSpec[]): Promise<void> {
+  utils.logBullet(clc.bold.cyan("functions: ") + "cleaning up build files...");
+  const gcrCleaner = new ContainerRegistryCleaner();
+  try {
+    await Promise.all(functions.map((func) => gcrCleaner.cleanupFunction(func)));
+  } catch (err) {
+    logger.debug("Failed to delete container registry artifacts with error", err);
+    utils.logLabeledWarning(
+      "functions",
+      "Unhnandled error cleaning up build files. This could result in a small monthly bill if not corrected"
+    );
+  }
+
+  // TODO: clean up Artifact Registry images as well.
+}
+
+export class ContainerRegistryCleaner {
+  readonly helpers: Record<string, ContainerRegistryHelper> = {};
+
+  private helper(location: string): ContainerRegistryHelper {
+    const subdomain = SUBDOMAIN_MAPPING[location] || "us";
+    if (!this.helpers[subdomain]) {
+      this.helpers[subdomain] = new ContainerRegistryHelper(subdomain);
+    }
+    return this.helpers[subdomain];
+  }
+
+  // GCFv1 has the directory structure:
+  // gcf/
+  //  +- <region>/
+  //        +- <uuid>
+  //             +- <hash> (tags: <FuncName>_version-<#>)
+  //             +- cache/ (Only present in first deploy of region)
+  //             |    +- <hash> (tags: latest)
+  //             +- worker/ (Only present in first deploy of region)
+  //                  +- <hash> (tags: latest)
+  //
+  // We'll parallel search for the valid <uuid> and their children
+  // until we find one with the right tag for the function name.
+  // The underlying Helper's caching should make this expensive for
+  // the first function and free for the next functions in the same
+  // region.
+  async cleanupFunction(func: backend.FunctionSpec): Promise<void> {
+    const helper = this.helper(func.region);
+    const uuids = (await helper.ls(`${func.project}/gcf/${func.region}`)).children;
+
+    const uuidTags: Record<string, string[]> = {};
+    const loadUuidTags: Promise<void>[] = [];
+    for (const uuid of uuids) {
+      loadUuidTags.push(
+        (async () => {
+          const path = `${func.project}/gcf/${func.region}/${uuid}`;
+          const tags = (await helper.ls(path)).tags;
+          uuidTags[path] = tags;
+        })()
+      );
+    }
+    await Promise.all(loadUuidTags);
+
+    const extractFunction = /^(.*)_version-\d+$/;
+    const entry = Object.entries(uuidTags).find(([, tags]) => {
```

Comment:
Since you throw away the keys in the .`find()` predicate function, should you use `Object.values()` instead of `Object.entries()` here?